File: ffi/ts/src/hashId.ts
import {
  calcKeyId,
  castToKeyId,
  deserializeKeyId,
  serializeKeyId,
} from './blsct'

import { ManagedObj, isWasmPtrWrapper, wrapWasmPtr } from './managedObj'
import { PublicKey } from './keys/publicKey'
import { Scalar } from './scalar'

/** Represents a hash ID consisting of a blinding public key, a spending public key, and a view key. Also known as `CKeyId` which is an alias for `uint160` on the C++ side.
 *
 * Examples:
 * ```ts
 * const { HashId, PublicKey, ChildKey, Scalar } = require('navio-blsct')
 * const blindingPubKey = PublicKey.random()
 * const spendingPubKey = PublicKey.random()
 * const seed = Scalar.random()
 * const viewKey = Scalar.random()
 * const hashId = HashId.generate(blindingPubKey, spendingPubKey, viewKey)
 * const ser = hashId.serialize()
 * const deser = HashId.deserialize(ser)
 * ser === deser.serialize() // true
 * ```
 */
export class HashId extends ManagedObj {
  /** Constructs a new `HashId` instance.
   * - If no parameter is provided, a random `HashId` is generated.
   * - If a WASM pointer wrapper is provided (from fromObj/deserialize), it wraps the pointer.
   */
  constructor(obj?: any) {
    if (isWasmPtrWrapper(obj)) {
      // WASM pointer from fromObj or _deserialize
      super(obj)
    } else if (typeof obj === 'object' && obj !== null) {
      // Native NAPI object
      super(obj)
    } else {
      const newObj = calcKeyId(
        PublicKey.random().value(),
        PublicKey.random().value(),
        Scalar.random().value(),
      )
      super(typeof newObj === 'number' ? wrapWasmPtr(newObj) : newObj)
    }
  }

  /** Generates a `HashId` from the provided blinding public key, spending public key, and view key.
   * @param blindingPubKey - The public key used for blinding.
   * @param spendingPubKey - The spending public key.
   * @param viewKey - The view key.
   * @returns A new `HashId` instance.
   */
  static generate(
    blindingPubKey: PublicKey,
    spendingPubKey: PublicKey,
    viewKey: Scalar,
  ): HashId {
    const obj = calcKeyId(
      blindingPubKey.value(),
      spendingPubKey.value(),
      viewKey.value()
    )
    return HashId.fromObj(obj)
  }

  /** Generates a random `HashId`.
   * @returns A new `HashId` instance with two random `PublicKey`s and a random `ViewKey`.
   */
  static random(): HashId {
    return HashId.generate(
      PublicKey.random(),
      PublicKey.random(),
      Scalar.random(),
    )
  }
  
  override value(): any {
    return castToKeyId(this.obj)
  }

  override serialize(): string {
    return serializeKeyId(this.value())
  }

  /** Deserializes a hexadecimal string into a `HashId` instance.
   * @param hex - The hexadecimal string to deserialize.
   * @returns A new `HashId` instance.
   */
  static deserialize(
    this: new (obj: any) => HashId,
    hex: string
  ): HashId {
    return HashId._deserialize(hex, deserializeKeyId)
  }
}

File: ffi/rust/src/hash_id.rs
use crate::{
  blsct_obj::{BlsctObj, self},
  blsct_serde::BlsctSerde, 
  ffi::{
    BlsctKeyId,
    BlsctRetVal,
    calc_key_id,
    deserialize_key_id,
    serialize_key_id,
  },
  keys::{
    child_key::ChildKey,
    public_key::PublicKey,
  },
  macros::{
    impl_clone,
    impl_display,
    impl_from_retval,
    impl_value,
  },
  scalar::Scalar,
};
use serde::{Deserialize, Serialize};
use std::ffi::c_char;

#[derive(Debug, Deserialize, Serialize, Eq)]
pub struct HashId {
  obj: BlsctObj<HashId, BlsctKeyId>,
}

impl_from_retval!(HashId);
impl_display!(HashId);
impl_clone!(HashId);

impl HashId {
  pub fn new(
    blinding_pub_key: &PublicKey,
    spending_pub_key: &PublicKey,
    view_key: &Scalar,
  ) -> Self {
    let blsct_key_id = unsafe { calc_key_id(
      blinding_pub_key.value(),
      spending_pub_key.value(),
      view_key.value(),
    )};
    BlsctObj::from_c_obj(blsct_key_id).into()
  }

  pub fn random<'a>() -> Result<Self, blsct_obj::Error<'a>> {
    let blinding_pub_key = PublicKey::random()?;
    let spending_pub_key = PublicKey::random()?;
    let view_key = {
      let child_key = ChildKey::random()?;
      child_key.to_tx_key().to_view_key() 
    };
    Ok(Self::new(
      &blinding_pub_key,
      &spending_pub_key,
      &view_key,
    ))
  }

  impl_value!(BlsctKeyId);
}

impl BlsctSerde for HashId {
  unsafe fn serialize(ptr: *const u8, _: usize) -> *const i8 {
    serialize_key_id(ptr as *const BlsctKeyId)
  }

  unsafe fn deserialize(hex: *const c_char) -> *mut BlsctRetVal {
    deserialize_key_id(hex)
  }
}

impl PartialEq for HashId {
  fn eq(&self, other: &Self) -> bool {
    self.obj == other.obj 
  }
}

impl From<BlsctObj<HashId, BlsctKeyId>> for HashId {
  fn from(obj: BlsctObj<HashId, BlsctKeyId>) -> HashId {
    HashId { obj }
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::initializer::init;

  #[test]
  fn test_eq() {
    init();
    let (a, b) = {
      loop {
        let a = HashId::random().unwrap();
        let b = HashId::random().unwrap();
        if a != b {
          break (a, b);
        }
      }
    };
    assert!(a == a);
    assert!(a != b);
    assert!(b != a);
    assert!(b == b);
  }

  #[test]
  fn test_deser() {
    init();
    let a = HashId::random().unwrap();
    let hex = bincode::serialize(&a).unwrap();
    let b = bincode::deserialize::<HashId>(&hex).unwrap();
    assert_eq!(a, b);
  }
}

